tools/rtpsrelay/RtpsRelay.cpp

```diff
@@ -155,13 +230,212 @@ int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
+    DDS::Topic_var topic = relay_participant->create_topic("Readers", type_name,
+                                                           TOPIC_QOS_DEFAULT, nullptr,
+                                                           OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+    if (!topic) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: failed to create Readers topic\n"));
+      return EXIT_FAILURE;
+    }
+
+    DDS::Publisher_var publisher = relay_participant->create_publisher(PUBLISHER_QOS_DEFAULT, nullptr,
+                                                                       OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+    if (!publisher) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: failed to create Reader publisher\n"));
+      return EXIT_FAILURE;
+    }
+
+    DDS::Subscriber_var subscriber = relay_participant->create_subscriber(SUBSCRIBER_QOS_DEFAULT, nullptr,
+                                                                          OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+    if (!subscriber) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: failed to create Reader subscriber\n"));
+      return EXIT_FAILURE;
+    }
+
+    DDS::DataWriterQos writer_qos;
+    publisher->get_default_datawriter_qos(writer_qos);
+
+    writer_qos.durability.kind = DDS::TRANSIENT_LOCAL_DURABILITY_QOS;
+
+    DDS::DataWriter_var writer = publisher->create_datawriter(topic, writer_qos, nullptr,
+                                                              OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+    if (!writer) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: failed to create Reader data writer\n"));
+      return EXIT_FAILURE;
+    }
+
+    RtpsRelay::ReaderEntryDataWriter_ptr reader_writer = RtpsRelay::ReaderEntryDataWriter::_narrow(writer);
+
+    if (!reader_writer) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: failed to narrow Reader data writer\n"));
+      return EXIT_FAILURE;
+    }
+
+    DDS::DataReaderQos reader_qos;
+    subscriber->get_default_datareader_qos(reader_qos);
+
+    reader_qos.durability.kind = DDS::TRANSIENT_LOCAL_DURABILITY_QOS;
+    reader_qos.reliability.kind = DDS::RELIABLE_RELIABILITY_QOS;
+
+    DDS::DataReader_var reader = subscriber->create_datareader(topic, reader_qos, nullptr,
+                                                               OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+    if (!reader) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: failed to create Reader data reader\n"));
+      return EXIT_FAILURE;
+    }
+
+    RtpsRelay::ReaderEntryDataReader_ptr reader_reader = RtpsRelay::ReaderEntryDataReader::_narrow(reader);
+
+    if (!reader_reader) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: failed to narrow Reader data reader\n"));
+      return EXIT_FAILURE;
+    }
+
+    DDS::DataReaderListener_var reader_listener(new ReaderListener(association_table));
+    DDS::ReturnCode_t ret = reader_reader->set_listener(reader_listener, OpenDDS::DCPS::DEFAULT_STATUS_MASK);
```

Comment:
For Entities this code is creating (not the Built-in ones): instead of `set_listener`, create the entity directly with the listener or turn off auto-enable.  Otherwise the application may be expecting events that it never gets.